feat(build): Add step to upload blobs from file based output

kitop opened the pull request:
🎉 Thanks for submitting a pull request! 🎉

#### Summary

This adds a step before deploy to upload blobs from a specific folder in your output directory (ie: `<dist>/.netlify/blobs/deploy/`) to the deploy-specific blob store.

This step only runs if there's a deploy id present and are files to upload. A deploy id will only be present during CI builds. For the CLI we'll have to call this separately (will do that in a follow-up PR).
To prepare for calling this from the CLI, I had to update a few attributes that we pass around, in particular:
- Add `quiet` all around and pass that to the core steps. This is needed since the CLI uses it.
- Document `deployId` in BuildFlags type
- Receive deployId, quiet, and token on the exposed `runCoreStep` method

As the last piece, this also adds some logic before the build command to clean up any existing files in the blob directory to make sure we only deploy blobs generated from by last build to the deploy store.

Part of https://linear.app/netlify/issue/ADN-88 and https://linear.app/netlify/issue/FRA-103

---

For us to review and ship your PR efficiently, please perform the following steps:

- [ ] Open a [bug/issue](https://github.com/netlify/build/issues/new/choose) before writing your code 🧑‍💻. This ensures
      we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or
      something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [ ] Read the [contribution guidelines](https://github.com/netlify/build/blob/main/CONTRIBUTING.md) 📖. This ensures
      your code follows our style guide and passes our tests.
- [ ] Update or add tests (if any source code was changed or added) 🧪
- [ ] Update or add documentation (if features were changed or added) 📝
- [ ] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**

JGAntunes reviewed (commented):
Haven't gone through the whole thing but leaving some comments early on. Will continue tomorrow 👍

kitop commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,54 @@
+      const { data, metadata } = await getFileWithMetadata(blobsDir, key)
+      await blobStore.set(key, data, { metadata })
+    }),
+  ).catch((err) => {
+    logError(logs, `Error uploading blobs to deploy store: ${err.message}`)
+
+    const error = new Error(`Failed while uploading blobs to deploy store.`)
+    addErrorInfo(error, { type: 'blobsUploadError' })
+    throw error
+  })
```
Yes! As soon as one upload fails, we wanna cancel everything and throw an error to fail the build, as we don't want to deploy if we couldn't upload everything to the blob store.

kitop commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,51 @@
+  const basename = path.basename(key)
+  const metadataPath = path.join(blobsDir, dirname, `${metadataPrefix}${basename}${metadataSuffix}`)
+
+  const data = fs.readFileSync(contentPath)
+  let metadata = {}
+  if (fs.existsSync(metadataPath)) {
+    const metadataFile = fs.readFileSync(metadataPath, { encoding: 'utf8' })
```
Good call! Will update.

kitop replied:
I've updated it in 3d21cdfa6 so it does async a bit better, however still using `existsSync` (hopefully it's a very fast check) since it's a bit nicer.
It'll probably benefit from using `fs.stat` (the promise version) and doing some extra checks though, right?

kitop commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,54 @@
+import { getKeysToUpload, getBlobsDir, getFileWithMetadata, anyBlobsToUpload } from './utils.js'
+
+const coreStep = async function ({ debug, logs, deployId, buildDir, constants: { PUBLISH_DIR, SITE_ID, TOKEN } }) {
+  const stepResults = {}
```
I thought I was going to return something, but ended up not doing so. I can clean up and always return an empty object.

kitop commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,54 @@
+  log(logs, `Uploading ${keys.length} blobs to deploy store...`)
+
+  await Promise.all(
+    keys.map(async (key) => {
+      if (debug) {
+        log(logs, `- Uploading blob ${key}`, { indent: true })
+      }
+      const { data, metadata } = await getFileWithMetadata(blobsDir, key)
+      await blobStore.set(key, data, { metadata })
+    }),
```
Oh, yes! I wasn't sure how to best approach this so went with a simple one.
We can probably use `pMap` like we do for uploading files in the CLI: https://github.com/netlify/cli/blob/main/src/utils/deploy/upload-files.ts#L74

What do you think?

eduardoboucas commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,62 @@
+  buildDir,
+  constants: { PUBLISH_DIR, SITE_ID, NETLIFY_API_TOKEN },
+}) {
+  const blobStore = getDeployStore({ siteID: SITE_ID, deployID: deployId, token: NETLIFY_API_TOKEN })
```
If we supply an `apiURL` property here (constructed with the value of `constants.NETLIFY_API_HOST`) we can make this work on staging.

kitop replied:
Nice catch! Done in 06f58d8a9

eduardoboucas commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,62 @@
+  const blobsDir = getBlobsDir({ buildDir, publishDir: PUBLISH_DIR })
+  const keys = await getKeysToUpload(blobsDir)
+
+  // We checked earlier, but lets be extra safe
```
```suggestion
  // We checked earlier, but let's be extra safe
```

eduardoboucas commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,61 @@
+
+import { fdir } from 'fdir'
+
+const metadataPrefix = '$'
+const metadataSuffix = '.json'
```
nit: Using capitalised names for static constants would be more in line with the rest of the codebase.

```suggestion
const METADATA_PREFIX = '$'
const METADATA_SUFFIX = '.json'
```

eduardoboucas commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,61 @@
+export const getBlobsDir = function ({ buildDir, publishDir }) {
+  const blobsPath = '.netlify/blobs/deploy'
+  const absoluteBlobsPath = path.resolve(buildDir, publishDir, blobsPath)
+  const blobsDir = path.normalize(absoluteBlobsPath)
```
Do we need `path.normalize` here considering we run `path.resolve` just above?

kitop replied:
Probably not, and it was my lack of knowledge on Node.js 😅 
Updated in 6ff3ab833

eduardoboucas commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,61 @@
+
+async function readMetadata(metadataPath: string): Promise<Record<string, string>> {
+  let metadata = {}
+  if (fs.existsSync(metadataPath)) {
```
Checking if the file exists before trying to read it is usually [not a recommended pattern](https://nodejs.org/api/fs.html#fsexistspath-callback). I would recommend doing the `readFile` in a `try/catch` and looking for `error.code === "ENOENT"` as signal that the file doesn't exist.

kitop replied:
Updated in f90332636

eduardoboucas commented on packages/build/tests/blobs_upload/tests.js:
```diff
@@ -0,0 +1,89 @@
+  const port = await getPort()
+  t.context.blobRequestCount = { set: 0, get: 0 }
+
+  t.context.blobServer = new BlobsServer({
```
Nice!

eduardoboucas commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,62 @@
+  coreStep,
+  coreStepId: 'blobs_upload',
+  coreStepName: 'Uploading blobs',
+  coreStepDescription: () => 'Uploading blobs to deploy store from .netlify/blobs/deploy output directory',
```
I'm a bit torn on the level of detail we should go into here. I wonder if saying things like "deploy store" and the internal blobs directory is confusing when most people won't need to interact with this type of low-level concepts, and maybe saying something like "Uploading blobs" would be less overwhelming?

Just a thought, and definitely not a blocker.

eduardoboucas reviewed (commented):
It's looking great! I left a few comments.

lukasholzer commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,61 @@
+
+export const anyBlobsToUpload = function ({ buildDir, publishDir }) {
+  const blobsDir = getBlobsDir({ buildDir, publishDir })
+  const { files } = new fdir().onlyCounts().crawl(blobsDir).sync()
```
shouldn't we use the promise based method here? I think it will be way faster as it can leverage async I/O

```suggestion
  const { files } = await new fdir().onlyCounts().crawl(blobsDir).withPromise()
```

lukasholzer replied:
Furthermore we are crawling the dir multiple times wouldn't make it sense to do this kind of cost-intensive operation once? 

I see you use it a second time on `getKeysToUpload`

kitop replied:
Updated to use the async version on 67684e2e3
I'm not too worried about doing the operation twice, the count should be pretty quick. If you're concerned we can memoize it but I'm less familiar with how to do that in this setup.

lukasholzer commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,61 @@
+  return files.map((f) => f.split(path.sep).join('/'))
+}
+
+// Read a file and its metadata from the blobs directory
```
```suggestion
/** Read a file and its metadata from the blobs directory **/
```

[dust]: for function annotations please use JSDoc as this can be picked up by the intelligence. a regular inline comment (`//`) wont show up in the intellisense

kitop replied:
Updated in d364a8dfe

kitop commented:
Thanks all for the review! I've updated everything based on the comments, and tried to address each individually. I'd love a second round of review if you don't mind. And in particular, I have a few questions blocking this from being ready:
- I'd like to do some cleanup of the blobs directory before starting the build. I wasn't sure what was the best place to do that. If I can hook it in the same plugin via a `onPreBuild` event maybe, or if put it in the `build_command` core step. Any thoughts or guidance are appreciated.
- Tests are failing in Node 14.16 due to the lack of `fetch`. I didn't find any fetch polyfill already, and given that the CLI has requirements for Node 18+ (see https://github.com/netlify/cli/blob/v17.6.0/package.json#L7-L9) I was wondering if it was worth doing so or we can drop the requirement for 14.16
- The tests are also failing on Windows, unexpectedly. Somehow doing more uploads (or requests to the BlobServer) than expected. Is this a known issue of netlify/blobs on Windows? Any pointers?
- Will leave a couple more comments in the code as well.
Thanks!

kitop commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,62 @@
+    logError(logs, `Error uploading blobs to deploy store: ${err.message}`)
+
+    const error = new Error(`Failed while uploading blobs to deploy store`)
+    addErrorInfo(error, { type: 'blobsUploadError' })
```
I wasn't sure about this and the`type` here. It's not defined in https://github.com/netlify/build/blob/4511447230ae5b582821b40499ae29d97af0aeae/packages/build/src/error/type.js#L42

Should something change? /cc @JGAntunes I think you're the expert in this area.

kitop replied:
Update after a chat: I'll remove the `addErrorInfo` and rely on the default `coreStep` added by the system.

kitop commented on packages/build/tests/blobs_upload/tests.js:
```diff
@@ -0,0 +1,89 @@
+  process.env.NETLIFY_BLOBS_CONTEXT = Buffer.from(
+    JSON.stringify({
+      edgeURL: `http://localhost:${port}`,
+      token: TOKEN, // This is for tests, during build this is passed as argument
```
I had to provide this here because providing `token` flag in the fixtures made the tests fail. I think it's because it tries to communicate with the API. I tried `runBuildServer` in the fixtures but didn't work. I'd love to test it properly with the token passed as argument rather than using this workaround for tests. Any pointers are appreciated!

eduardoboucas replied:
Are you saying you tried to pass `token` as an option to `getDeployStore` and that failed? That's strange. Happy to pair on this.

kitop replied:
No, sorry. I tried passing it to `withFlags` on the tests so it gets passed automatically, but it failed. So in https://github.com/netlify/build/pull/5404/files#diff-05505b9d08762ab76b5e42a735b1b91aeb802a3ddd475f25002474e5e593c763R56-R59

I tried this and tests stalled
```
const output = await new Fixture('./fixtures/src_with_blobs')
    .withFlags({ deployId: 'abc123', siteId: 'test', token: TOKEN })
    .runWithBuild()
```

eduardoboucas commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  try {
+    await pMap(keys, uploadBlob, { concurrency: 10 })
```
FYI, the concurrency limit we currently use in the CLI is 5: https://github.com/netlify/cli/blob/5c61169d43570ffb20cfd2976ca050a9558d9c3f/src/utils/deploy/constants.ts#L6.

I don't have any thoughts on which value we should use, just wanted to surface what we're doing in the CLI.

kitop replied:
Oh good catch. From buildbot we do 10, so I think I had that one in mind. I think we should be OK keeping it at 10 (it's a much simpler upload than a file upload), but I was planning on keeping an eye on buildbot traces anyway.

eduardoboucas commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,71 @@
+const METADATA_SUFFIX = '.json'
+
+export const getBlobsDir = function ({ buildDir, publishDir }) {
+  const blobsPath = '.netlify/blobs/deploy'
```
nit: You could move this to a `BLOBS_PATH` constant in the global scope.

eduardoboucas commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,71 @@
+    }
+    throw err
+  }
+  let metadata = {}
+  try {
+    metadata = JSON.parse(metadataFile)
+  } catch (err) {
+    // Normalize the error message
+    throw new Error(`Error parsing metadata file '${metadataPath}'`)
+  }
+  return metadata
```
nit: You could simplify this a bit, since you're either returning the value of `JSON.parse` or throwing.

```suggestion
  try {
    return JSON.parse(metadataFile)
  } catch (err) {
    // Normalize the error message
    throw new Error(`Error parsing metadata file '${metadataPath}'`)
  }
```

eduardoboucas commented on packages/build/tests/blobs_upload/tests.js:
```diff
@@ -0,0 +1,89 @@
+
+  const store = getDeployStore({ deployID: 'abc123', siteID: 'test' })
+
+  let blob
+  blob = await store.getWithMetadata('something.txt')
+  t.is(blob.data, 'some value')
+  t.deepEqual(blob.metadata, {})
+
+  blob = await store.getWithMetadata('with-metadata.txt')
+  t.is(blob.data, 'another value')
+  t.deepEqual(blob.metadata, { meta: 'data', number: 1234 })
+
+  blob = await store.getWithMetadata('nested/file.txt')
+  t.is(blob.data, 'file value')
+  t.deepEqual(blob.metadata, { some: 'metadata' })
```
nit: I would avoid reassigning if we can avoid it. This boils down to preference, and is especially inconsequential in a test, so by all means feel free to take or leave.

```suggestion
const blob1 = await store.getWithMetadata('something.txt')
t.is(blob1.data, 'some value')
t.deepEqual(blob1.metadata, {})

const blob2 = await store.getWithMetadata('with-metadata.txt')
t.is(blob2.data, 'another value')
t.deepEqual(blob2.metadata, { meta: 'data', number: 1234 })

const blob3 = await store.getWithMetadata('nested/file.txt')
t.is(blob3.data, 'file value')
t.deepEqual(blob3.metadata, { some: 'metadata' })
```

eduardoboucas reviewed (approved):
LGTM!

eduardoboucas commented:
> * I'd like to do some cleanup of the blobs directory before starting the build. I wasn't sure what was the best place to do that. If I can hook it in the same plugin via a `onPreBuild` event maybe, or if put it in the `build_command` core step. Any thoughts or guidance are appreciated.

I think doing it in the same plugin as part of `onPreBuild` would be nice because then all the logic would be in the same file.



> * Tests are failing in Node 14.16 due to the lack of `fetch`. I didn't find any fetch polyfill already, and given that the CLI has requirements for Node 18+ (see https://github.com/netlify/cli/blob/v17.6.0/package.json#L7-L9) I was wondering if it was worth doing so or we can drop the requirement for 14.16

I think it would make sense to drop the requirement for Node 14 (and we'll need to do it for Node 16 too, since `fetch` is 18+ only). I always get tripped up by the versions we need to support, though, so I'd love to hear what others think.



> * The tests are also failing on Windows, unexpectedly. Somehow doing more uploads (or requests to the BlobServer) than expected. Is this a known issue of netlify/blobs on Windows? Any pointers?

No issues that I know of. I'm happy to pair with you on this to troubleshoot further, if you want.

kitop commented:
> > * I'd like to do some cleanup of the blobs directory before starting the build. I wasn't sure what was the best place to do that. If I can hook it in the same plugin via a `onPreBuild` event maybe, or if put it in the `build_command` core step. Any thoughts or guidance are appreciated.
> 
> I think doing it in the same plugin as part of `onPreBuild` would be nice because then all the logic would be in the same file.

Thank you! Does I couldn't find examples of a corePlugin having 2+ events. Do I need to add another function and tie it to `event`? Or add it to the list [here](https://github.com/netlify/build/pull/5404/files#diff-c0b115b1e11cffc188480fa1c709ee4e7522ea576396d900424856d838d6d8bcR70-R79)?
Any pointers are appreciated!

 
> > * Tests are failing in Node 14.16 due to the lack of `fetch`. I didn't find any fetch polyfill already, and given that the CLI has requirements for Node 18+ (see https://github.com/netlify/cli/blob/v17.6.0/package.json#L7-L9) I was wondering if it was worth doing so or we can drop the requirement for 14.16
> 
> I think it would make sense to drop the requirement for Node 14 (and we'll need to do it for Node 16 too, since `fetch` is 18+ only). I always get tripped up by the versions we need to support, though, so I'd love to hear what others think.

Oh, forgot it was only on Node 18+. Our build system currently runs Node 18, so dropping this shouldn't be terrible. However I might be missing something. I'd love to hear from @JGAntunes and @lukasholzer. If it becomes a blocker, we can add a fetch polyfill and remove once we're on Node 18.

> > * The tests are also failing on Windows, unexpectedly. Somehow doing more uploads (or requests to the BlobServer) than expected. Is this a known issue of netlify/blobs on Windows? Any pointers?
> 
> No issues that I know of. I'm happy to pair with you on this to troubleshoot further, if you want.

Thank you! I'll ping you later since I'm a bit new in this codebase and setup.

kitop commented:
I've paired a bit with Eduardo. Some updates on the tests:
- I've added polyfill for Node < 18. We can discuss removing Node 16 support at a separate point to not block this.
- I've added the token in the test, however I had to use a workaround by passing `offline` to not ask for site config.
- I tried updating the Windows tests, but didn't get too far. Some debug logging helps and the error seems to be when writing the files inside the blob server. I updated the blob server setup to be the same as in [netlify/blob tests](https://github.com/netlify/blobs/blob/main/src/server.test.ts#L44-L49). I'm a bit confused at what can be the issue here. It's taking me a bit of time to try to recreate this locally.

The only thing pending is adding some cleanup phase before the build. We can't have `onPreBuild` in coreSteps so I'll see if it's worth adding a new step or putting it as part of the main build one.

kitop commented:
Ohhh I found the windows error!
We can't use `:` in a file name in windows!
<img width="834" alt="CleanShot 2023-11-23 at 17 03 58@2x" src="https://github.com/netlify/build/assets/143179/fcc877c1-f4db-410f-b8d6-fff006a5ce36">
<img width="823" alt="CleanShot 2023-11-23 at 17 05 01@2x" src="https://github.com/netlify/build/assets/143179/e0e182f0-f2df-432c-b648-1b7b9810ae19">

We might have to do a workaround in netlify/blobs for it! /cc @eduardoboucas

eduardoboucas commented:
Ah, great catch! Do you mind opening an issue in Linear and assigning to me? I can look into that tomorrow.

In the meantime, I think it's fine if you skip the test on Windows so you're not blocked.

kitop commented:
Added something in the build command step to do a bit of cleanup before starting. We talked with Eduardo about maybe doing a real "cleanup core step", but this would mean it will log stuff, and the lift needed to make it so we don't log is a bit over what I wanted to tackle. We can start this way, and then cleanup if we add that capability.
However, I wasn't sure how/where to test that, any pointers appreciated!

This and the window thing are the missing 2 pieces for this to be ready. Thanks all for the help so far!!

kitop commented:
Ok, this is ready for final review! Sorry for all the noise with the force pushes, I tried to clean up the git history a bit so it's also easier to read commit by commit if you want to.

Added some notes to the PR description as well. I have a CLI PR almost ready for when these changes go live.

lukasholzer commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,84 @@
+    token: NETLIFY_API_TOKEN,
+    apiURL: API_URL,
+  }
+  if (semver.lt(nodeVersion, '18.0.0')) {
```
nice!

kitop replied:
We will be calling this step from the CLI when doing `netlify deploy` (with or without `--build`).
However this version check is needed because we still run tests with Node 14 in CI here because we still support Node 14/16 https://github.com/netlify/build/blob/main/packages/build/package.json#L155
CLI already requires Node 18+, and in buildbot we run Node 18 as well for this piece. However, I didn't want to block this work on updating requirements for build. We can do that as a follow-up.

lukasholzer commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,66 @@
+  return files.map((f) => f.split(path.sep).join('/'))
+}
+
+/* Read a file and its metadata file from the blobs directory */
```
```suggestion
/** Read a file and its metadata file from the blobs directory */
```

A valid JsDoc comment needs to start with two asterisks. Otherwise the intellisense cannot pick it up 😬

lukasholzer commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,66 @@
+  return files > 0
+}
+
+/* Given output directory, find all file paths to upload excluding metadata files */
```
```suggestion
/** Given output directory, find all file paths to upload excluding metadata files */
```

lukasholzer commented on packages/build/tests/blobs_upload/tests.js:
```diff
@@ -0,0 +1,99 @@
+  t.snapshot(normalizeOutput(output))
+})
+
+test.serial('blobs upload, uploads files to deploy store', async (t) => {
```
is there a reason why those have to be serial in comparison to the other two on the top?

I think as you get a unique port for each test all can be concurrently. (the default)

kitop replied:
I was running into flaky tests when not running it serially. But it might have been something else. I'll change and we can always add it back if needed.

lukasholzer commented on packages/build/src/plugins_core/build_command.js:
```diff
@@ -6,6 +7,16 @@ import { addErrorInfo } from '../error/info.js'
+  const blobsDir = getBlobsDir({ buildDir, publishDir })
+  try {
+    await fsPromises.rm(blobsDir, { recursive: true, force: true })
+  } catch (err) {
```
```suggestion
  } catch {
```

lukasholzer commented on packages/build/src/plugins_core/build_command.js:
```diff
@@ -1,3 +1,4 @@
+import fsPromises from 'node:fs/promises'
```
```suggestion
import { rm } from 'node:fs/promises'
```
[dust]: always prefer named import over default imports

lukasholzer commented on packages/build/src/plugins_core/build_command.js:
```diff
@@ -6,6 +7,16 @@ import { addErrorInfo } from '../error/info.js'
+async function cleanupBlobsDir(buildDir, publishDir) {
+  const blobsDir = getBlobsDir({ buildDir, publishDir })
+  try {
+    await fsPromises.rm(blobsDir, { recursive: true, force: true })
```
```suggestion
    await rm(blobsDir, { recursive: true, force: true })
```

[dust]: use the function from the named import

lukasholzer commented on packages/build/src/plugins_core/blobs_upload/utils.ts:
```diff
@@ -0,0 +1,66 @@
+
+  try {
+    return JSON.parse(metadataFile)
+  } catch (err) {
```
```suggestion
  } catch  {
```

[dust]: you can omit the error if you don't need it.

lukasholzer reviewed (approved):
Awesome work! This is huge 🚀 

Left some minor comments nothing blocking though.

JGAntunes commented:
About this:

> Oh, forgot it was only on Node 18+. Our build system currently runs Node 18, so dropping this shouldn't be terrible. However I might be missing something. I'd love to hear from @JGAntunes and @lukasholzer. If it becomes a blocker, we can add a fetch polyfill and remove once we're on Node 18.

My understanding is that these core steps will always execute with the system node version so, in terms of our CI/CD system, we should be fine (since we're running 18.x currently). For the CLI this core step won't run right(?) since we'll be running the custom CLI logic required because of the "build before deploy logic"? tl;dr I think we only need to support node 18+?

JGAntunes commented on packages/build/tests/blobs_upload/tests.js:
```diff
@@ -0,0 +1,99 @@
+  t.snapshot(normalizeOutput(output))
+})
+
+test.serial('blobs upload, cancels deploy if blob metadata is malformed', async (t) => {
+  const output = await new Fixture('./fixtures/src_with_malformed_blobs_metadata')
+    .withFlags({ deployId: 'abc123', siteId: 'test', token: TOKEN, offline: true })
+    .runWithBuild()
+
+  t.is(t.context.blobRequestCount.set, 0)
+
+  t.snapshot(normalizeOutput(output))
+})
```
Thoughts on making this a programmatic test VS snapshot one where we can assert on the build exit code, severity, etc.?

- https://github.com/netlify/build/blob/5eaf9de39853307607b8882ba7a47aef9aa083bf/packages/testing/src/fixture.ts#L196-L199
- https://github.com/netlify/build/blob/5eaf9de39853307607b8882ba7a47aef9aa083bf/packages/build/tests/secrets_scanning/tests.js#L62-L72

kitop replied:
Ohhh nice! Yes!

JGAntunes commented on packages/build/tests/blobs_upload/tests.js:
```diff
@@ -0,0 +1,99 @@
+  t.is(blob3.data, 'file value')
+  t.deepEqual(blob3.metadata, { some: 'metadata' })
+
+  t.snapshot(normalizeOutput(output))
```
Wondering if we still need a full snapshot test or if we could do some partial assertion on the output and that would be enough? Not a blocker.

JGAntunes commented on packages/build/tests/blobs_upload/tests.js:
```diff
@@ -0,0 +1,99 @@
+
+  t.is(t.context.blobRequestCount.set, 0)
+
+  t.snapshot(normalizeOutput(output))
```
Wondering if we still need a full snapshot test or if we could do some partial assertion on the output and that would be enough? (check there's nothing referencing blob upload or something). Not a blocker.

kitop replied:
I think we can try to assert on presence or lack of presence of messages. I'll give it a go

JGAntunes commented on packages/build/tests/steps/tests.js:
```diff
@@ -41,3 +43,19 @@ test('Invalid package.json does not make build fail', async (t) => {
+
+  const blobsDir = path.join(fixture.repositoryRoot, 'out', '.netlify', 'blobs', 'deploy')
+
+  await t.throwsAsync(fsPromises.access(blobsDir))
```
Should we do this same check before the test execution (validating the dir exists)? Just to make sure we fully test the behaviour we want, having been bitten by stuff like this in the past 😅

kitop replied:
💯

JGAntunes commented on packages/build/src/plugins_core/build_command.js:
```diff
@@ -14,10 +25,12 @@ const coreStep = async function ({
     build: { command: buildCommand, commandOrigin: buildCommandOrigin },
   },
 }) {
+  await cleanupBlobsDir(buildDir, PUBLISH_DIR)
```
If I'm understanding this correctly, this effectively means we're saying that no steps before `onBuild` can create blobs since these will be removed, are we ok with that?

kitop replied:
Yeap. I wasn't quite sure of where to put it, to be honest. An `onPreBuild` or something that runs very very first would be ideal, but wasn't sure how to do that in a clean way. Any pointers appreciated!

kitop replied:
Ping @eduardoboucas here as well. Not sure how big of an issue it is (we can document it if it goes live like this maybe?).

JGAntunes replied:
> Yeap. I wasn't quite sure of where to put it, to be honest. An onPreBuild or something that runs very very first would be ideal, but wasn't sure how to do that in a clean way. Any pointers appreciated!

I'm thinking probably the cleanest way is to have a `onPreBuild` core step that runs before everything else and where we can do any cleanup (this is one example, but we could potentially use it for anything else 🤷). Happy to leave that for a further PR.

eduardoboucas replied:
> I'm thinking probably the cleanest way is to have a `onPreBuild` core step that runs before everything else and where we can do any cleanup (this is one example, but we could potentially use it for anything else 🤷).

I agree!

kitop replied:
I have a draft PR for the changes: https://github.com/netlify/build/pull/5413
If OK, I can push those here, otherwise we can merge this one and keep iterating on that one.

eduardoboucas replied:
Another side-effect of the current approach is that we won't do the cleanup if your site doesn't have a build command, right? I know it's a bit of an edge case, but I think it means we could end up pushing blobs from a previous build?

That draft PR looks good to me. If you want to push those changes here, go for it!

JGAntunes replied:
> That draft PR looks good to me. If you want to push those changes here, go for it!

Agree 👍 also happy to have it as a follow up, however y'all prefer.

> Another side-effect of the current approach is that we won't do the cleanup if your site doesn't have a build command, right?

Ah right!! Good point 👍

kitop replied:
> Another side-effect of the current approach is that we won't do the cleanup if your site doesn't have a build command, right?

Correct. Not ideal, agree.

I'll merge this to get things going and prep the other PR as a fast-follow.

JGAntunes reviewed (commented):
Left a couple of non-blocking comments. Mostly questions, tiny suggestions and dire wish to move away from snapshot testing 🙃

kitop commented on packages/build/package.json:
```diff
@@ -95,11 +96,13 @@
     "log-process-errors": "^8.0.0",
     "map-obj": "^5.0.0",
     "memoize-one": "^6.0.0",
+    "node-fetch": "^3.3.2",
```
Oh, nice! Is it full web-platform `fetch` compatible? We only need it because of that (see https://github.com/netlify/blobs?tab=readme-ov-file#custom-fetch)
From my understanding it'll need https://github.com/alexghr/got-fetch.

Would love to not add one more dependency though, but as mentioned in another piece we can probably remove this if we move build requirements to be Node 18+ (which we already do in CLI and it's the version we run in buildbot)

kitop replied:
Ignoring this for now as `got` didn't quite work and we need to add a dependency regardless. We should be cleaning this up somewhat soon and stop supporting Node <18 in here anyway.

kitop commented on packages/build/src/plugins_core/blobs_upload/index.ts:
```diff
@@ -0,0 +1,84 @@
+
+  // We checked earlier, but let's be extra safe
+  if (keys.length === 0) {
+    if (!quiet) {
```
> is this because we'll call this "manually" from the CLI and we want to make sure we don't log anything in that case?

Correct, in case we call it with `--json` or similar flags (in fact, I found out about this when working on the CLI tests that used `--json`)

> If so and this is a tiny little improvement, making this a wrapper that moves away the constant quiet check could be a nice + for readibility, what do you think?

I thought about that and asn't sure though if to make it in this package or in the general logger (in fact, I was surprised `log` didn't take into account the `quiet` check). I'll start by making it in this package only, but happy to change to a more global one.

kitop replied:
Will leave this for a follow-up as we have something similar in other places already and I prefer consistency, and don't wanna increase the scope of this.

eduardoboucas reviewed (approved):
🚀
